### PR TITLE
feat: add form constraints story

### PR DESCRIPTION
**Pull request by kchia**

### CODE REVIEW

**Comment by kchia** [comment]
My hunch is that for the average student, this user story is probably going to take at least 2 days.

File: front-end/src/reservations/ReservationCreate.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ function DishCreate() {
     const abortController = new AbortController();
     setError(null);
     createReservation(dish, abortController.signal)
-      .then((savedReservation) => {
+      .then(savedReservation => {
```
Comment (dalelotts) [inline front-end/src/reservations/ReservationCreate.js:17]: This change violates the Prettier coding standard and will get reverted as soon as someone runs prettier on this code.

File: front-end/src/utils/generateInvalidTime.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      time =
+        convertISOTimeToMinutes(RESTAURANT_SCHEDULE.closeTime) -
+        RESTAURANT_SCHEDULE.closeTimeBuffer +
+        offset;
+      break;
+    default:
+      break;
+  }
+
+  let hour = Math.floor(time / 60);
```
Comment (dalelotts) [inline front-end/src/utils/generateInvalidTime.js:29]: should be const

File: front-end/src/utils/generateInvalidTime.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  let minutes = time % 60;
```
Comment (dalelotts) [inline front-end/src/utils/generateInvalidTime.js:30]: should be const

File: front-end/src/utils/generateInvalidTime.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import convertISOTimeToMinutes from "./convertISOTimeToMinutes";
+import { RESTAURANT_SCHEDULE } from "../reservations/constants";
```
Comment (dalelotts) [inline front-end/src/utils/generateInvalidTime.js:2]: This file should be in `__test__/utils` and comments that it is for test use only. If you give the student the desired open and close time there is no need for this code.

File: front-end/src/utils/generateNextDate.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+const { RESTAURANT_SCHEDULE } = require("../reservations/constants");
+
+export default function generateNextDate(working = true) {
```
Comment (dalelotts) [inline front-end/src/utils/generateNextDate.js:3]: same as above

File: front-end/src/utils/generateValidTime.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import convertISOTimeToMinutes from "./convertISOTimeToMinutes";
+import { RESTAURANT_SCHEDULE } from "../reservations/constants";
+
```
Comment (dalelotts) [inline front-end/src/utils/generateValidTime.js:3]: same as above

**Review by dalelotts — CHANGES_REQUESTED** [review]
Just a few minor changes. The backend validation is missing entirely so the API will still accept invalid reservations. As always, both frontend and backend validation much exist.

**Comment by dalelotts** [comment]
I'll also add that the frontend validation can reasonably be  limited restricting input to open dates/hours only - which will be much easier. make use of min/max attributes rather than writing code. You can delegate to the backend for validation that requires code.

File: README.md [hunk 7]
```diff
@@ -77,4 +76,16 @@ so that I know how many customers will arrive at the restaurant on a given day.
 >
 > You do no need to worry about different or changing time zones for the dates or times.
 
-### US-2 TBD
+### US-2 Add reservation form validation
+
+As a restaurant manager
+I want to add constraints on the reservation
+so that my customers can only make a reservation during eligible days and times
```
Comment (dalelotts) [inline README.md:83]: I want to prevent users from creating reservations before we open or within XX minutes before closing

**Review by dalelotts — CHANGES_REQUESTED** [review]

**Comment by dalelotts** [comment]
we need to have the ability to run unit and e2e tests so the e2e tests are in the e2e directory.
 
 Can you also restore the App test?

**Comment by kchia** [comment]
@dalelotts what is App.test needed for?

File: front-end/src/utils/generateInvalidTime.js (hunk 4 above)
Comment (kchia) [reply front-end/src/utils/generateInvalidTime.js]: Would that be a bit too "constraining"?

File: README.md (hunk 7 above)
Comment (kchia) [reply README.md:83]: Hmm...the example provided above does not adequately describe the user story, so I'm going to stick with my original version with some minor updates.

File: front-end/package.json [hunk 8]
```diff
@@ -25,7 +26,7 @@
     "start:frontend": "npx cross-env BROWSER=none NODE_ENV=test react-scripts start",
     "build": "react-scripts build",
     "test": "npx cross-env CI=true run-s test:unit test:e2e",
-    "test:unit": "react-scripts test ",
+    "test:unit": "react-scripts test --testPathIgnorePatterns=src/__tests__/utils",
```
Comment (dalelotts) [inline front-end/package.json:29]: Is there some reason these cannot be moved to /e2e?

**Comment by kchia** [comment]
Moving these as we speak.
ᐧ

File: README.md [hunk 9]
```diff
@@ -97,4 +96,29 @@ so that I know how many customers will arrive at the restaurant on a given day.
...
+
+#### Acceptance criteria
+
+Given that the restaurant mananger is on the `/reservations/new` page,
+When the restaurant manager submits a reservation form,
+Display an error message if any of the following constraints are violated:
+
+- The date and time of the reservation must occur some time in the future, on a day the restaurant is open for service (the restaurant is closed on Tuesdays).
+
+### US-3 Create reservation within eligible timeframe
```
Comment (kchia) [inline README.md:112]: @dalelotts I ultimately ended up separating the user stories vertically instead...

File: front-end/package.json [hunk 10]
```diff
@@ -25,7 +26,7 @@
     "start:frontend": "npx cross-env BROWSER=none NODE_ENV=test react-scripts start",
     "build": "react-scripts build",
     "test": "npx cross-env CI=true run-s test:unit test:e2e",
-    "test:unit": "react-scripts test ",
+    "test:unit": "react-scripts test --testPathIgnorePatterns=e2e/utils",
```
Comment (dalelotts) [inline front-end/package.json:29]: Is this needed? I don't think /e2e is in the default search path.

Comment (kchia) [reply front-end/package.json]: Not needed anymore.

**Pull request merged**